main/input_default.cpp

```diff
@@ -135,6 +135,18 @@ float InputDefault::get_action_strength(const StringName &p_action) const {
 	return E->get().strength;
 }
 
+float InputDefault::get_action_duration(const StringName &p_action) const {
+	const Map<StringName, Action>::Element *E = action_state.find(p_action);
+	if (!E)
+		return 0.0f;
+
+	if (Engine::get_singleton()->is_in_physics_frame()) {
```

Comment:
If this gets accepted anyway, this wont work. The number of frames per second in a game might vary, so this computation is definitely not going to output a correct duration if the FPS fluctuate. The timestamp of the action state change should be stored instead to compute the duration.

Eventually, this function could be replaced by a `get_toggle_timestamp()` instead, to retrieve this timestamp directly instead of the duration.

Reply:
Currently the `Engine` singleton provides the following methods related to FPS:
```cpp
Engine::get_iterations_per_second()
Engine::get_frames_per_second()
// Engine::get_target_fps()
```
From my understanding, `Engine::get_iterations_per_second()` returns the fixed/physics FPS.
If you look below, the idle frame timestamp is used in conjunction with `Engine::get_frames_per_second()` which can fluctuate as you say.

So if we take the fixed FPS, this should output the right duration, am I right?

If we take the idle FPS, then it should compute the average time instead? If that's the case then it should be documented then, because in most cases we're interested in fixed process for logic. That's why I linked #26828 because the engine already seem to have this kind of an issue in the first place.

> The timestamp of the action state change should be stored instead to compute the duration.

The engine already does it, unless you say about exposing the timestamp to script.

Doing this via script won't help either even with `get_toggle_timestamp()` added because the user will still try to implement similar logic for computing duration as in here.

Reply:
I don't get why you would need all of this, just compute `current_time - time_the_action_changed_state`, that's a lot simpler than trying to use `Engine::get_iterations_per_second()` or `Engine::get_frames_per_second()`.

> So if we take the fixed FPS, this should output the right duration, am I right?

That's not even sure, it might happen the physics frame gets delayed if there are lags. Maybe dropped too, I am not sure.

> The engine already does it, unless you say about exposing the timestamp to script.

It does not look like so. See the `Action` struct in `input_default.h` :
```
	struct Action {
		uint64_t physics_frame;
		uint64_t idle_frame;
		bool pressed;
		float strength;
	};
```
You would just have to store the last update timestamp there, and compute the difference.

Reply:
Ah I think I get what you mean. You say that the method should just return the difference in frames rather than seconds? That's actually was my initial idea but I thought that it's not as user-friendly as returning the time in seconds, not sure about that.

> It does not look like so. See the Action struct in input_default.h

That's was my first impression too, until I stumbled upon this:

https://github.com/godotengine/godot/blob/dace1ec9126c752792a0566cfd9485a8ad40c0d1/main/input_default.cpp#L425-L440

That's the only part which seems to populate input action states on event parsing. Notice this line of code:
```cpp
is_action_pressed(E->key()) != p_event->is_action_pressed(E->key())
```
It means that the action's state is only updated on change.

So the `Action` struct should actually be more explicit regarding timestamp names (notice `last_*` prefix):
```cpp
struct Action {
	uint64_t last_physics_frame;
	uint64_t last_idle_frame;
	bool pressed;
	float strength;
};
```
And by the way, the `action_state` map is essentially empty if no actions been pressed during the game, so `Input.get_action_duration()` would not actually report the "unpressed" time but `0.0`. I wonder whether it makes sense to "pre-populate" the `action_state` from the `InputMap` on `Input` initialization to help this.

Reply:
> Ah I think I get what you mean. You say that the method should just return the difference in frames rather than seconds?

No, not at all. I just meant you need to store the timestamp to compute the duration when the user asks it.

> It means that the action's state is only updated on change.

That's why you need to store the timestamp of the last change.

> So the Action struct should actually be more explicit regarding timestamp names.

Those are not timestamps, they correspond to the number of frames/number of physics ticks ran since the beginning of the game. A timestamp should be added as a new field.

Reply:
I suppose you're talking about using `OS.get_ticks_usec` for timestamps then. But what if a game lags so much that the tick difference gets too big for a single frame?

Here are the changes to the method:
```cpp
float InputDefault::get_action_duration(const StringName &p_action) const {
	const Map<StringName, Action>::Element *E = action_state.find(p_action);
	if (!E)
		return 0;
	return (OS::get_singleton()->get_ticks_usec() - E->get().timestamp) / 1000000.0f;
}
```

`is_equal_approx(Input.get_action_duration("jump"), 0.0)` never reports `True` with this.

So I suppose the current changes are more robust even with possible issues with fluctuating FPS in my opinion.

Reply:
``get_ticks_usec() is definitely more robust than what you are using right now. Having a peak in a frame is quite rare, and the difference it might cause is at worst the duration of the frame itself, which doesn't even matter. You're solution cannot be accepted in it's current state, as any drop in FPS is going to make the computation completely wrong.

Using is_equal_approx(Input.get_action_duration("jump"), 0.0)` does not make a lot of sense anyway. That's what `is_action_just_pressed() is for.

Reply:
@groud the only question I have left is: when I push those changes, are you going to approve the PR personally? Yes or no. 🙂

Reply:
Well, not really. I am still against adding more such logic there, but other contributors might think it's a good idea. What I asked here is only a requirement on the implementation.